### DISCUSSION TITLE
More configuration options for gestures

### DISCUSSION THREAD

**Discussion by mister-drgn** [linuxmint/discussions#102] [Ideas]
I appreciate that Cinnamon 5.8 now uses touchegg for gesture support. However, it doesn't provide the full set of configuration options touchegg uses. Notably (unless I'm missing an option somewhere), Cinnamon's gesture actions can only trigger after you release your fingers from the touchpad. Imho, this makes gesture actions appear sluggish. In my own touchegg configuration, I have gesture actions set to trigger as soon as the gesture is detected:

`<on>begin</on>`

Is there any plan to add more gesture configuration options? Or is there a workaround for the current setup?

Thank you.

**Comment by Starcross** [comment]
I would be great to have more options for the actions as well. Defining a keyboard shortcut to be run is particularly useful.

In addition to the perceived sluggishness, I'm also finding my gestures are 'missed' much more often compared to using touchegg in previous versions of Cinnamon

**Comment by mtwebster** [comment]
Touchegg has a daemon and a client as separate processes.

The touchegg 'daemon' process (the part that interacts with hardware) is *exactly* the same touchegg daemon you were using previously - whichever version is available from the PPA (or was at the time of the Mint 21.2 release). Other than potentially false positive/negative device detection, any *hardware* issues are more than likely not caused by Cinnamon.

The client process is what Cinnamon replaces now. This is what receives events from the daemon and interprets them as whatever actions we've defined. I'll try to get keystrokes added for the next version, though you can work around this limitation if you're determined, using the custom command option.

As far as I know regarding sluggishness, the actions themselves are not being triggered earlier in the official touchegg client, you're just getting visual cues when a gesture starts (like beginning a workspace switch shows a little tab on the side of the screen).  The action still doesn't happen until the gesture completes (at least this was my observation). Eventually we plan to have 1-to-1 gestures (like when you swipe your phone or tablet, the page moves as you swipe).

I'm sure our device detection isn't perfect - some devices may not be detected when they should be. It's not something I can ask the daemon about currently. I'll add an override for people that 'know' they have a supported device, even though Cinnamon disagrees, and we'll try to improve automatic detection in the future.

Feel free to add more to this discussion - requests, etc... I just wanted to respond and clear up confusion about what's happening here.
Thanks

**Reply by grepppo** [reply]
Whats the best way of getting these overrides added? 

I have an HP Elite X2 which the touchegg daemon correctly detects the touchscreen (not the touchpad, but that's fine) but the Gestures App says that there are no compatible devices.

But when I manually set up a touchegg config file and run the touchegg client in a console it all works fine.

The full history on this is on the Mint forums here: https://forums.linuxmint.com/viewtopic.php?t=401304&start=20

**Reply by igarca** [reply]
I'd like to support the idea of gesture actions being triggered at the beginning of the movement and not at the end. Also, I would like to suggest that the pinch configuration includes "zoom" (to do pinch-to-zoom). I know about the Zoom Accessibility option, but it doesn't work smoothly. In addition to not being able to use it with the touchpad, the cursor does not remain motionless. When you zoom in or out with the touchpad, you're also scrolling the page, which makes the operation clunky.
Thank you!!!

**Comment by mister-drgn** [comment]
@Starcross A couple quick thoughts:
1) You can do keyboard shortcuts now by selecting custom command and using xdotool. An example for moving to the next workspace would be:
`xdotool key ctrl+alt+Right`
2) Have you tried adjusting the detection threshold for swipe gestures? That's available in Cinnamon 5.8's current gesture setup if you go to the rightmost section. I had to set it to the lowest possible value to detect gestures as well as I get using touchegg directly. Imho, the default value is too high.

@mtwebster Thanks for weighing in. Just to be clear: On my current touchegg setup, if I swipe three fingers to the right, I get an immediate switch to the next workspace. On Cinnamon 5.8, if I swipe to the right, nothing happens until I lift my fingers. It's not an issue of speed--it's an issue of whether the action triggers when I begin the gesture or when I finish the gesture. This is a big difference, which is configurable I believe with `<on>...</on> `in touchegg's config file. I think making this customizable or changing the default behavior in Cinnamon 5.8+ would be a big advantage.

One-to-one gestures sounds very cool. I imagine that would require going beyond touchegg?

**Reply by Starcross** [reply]
Thanks for this! 

1. Using `xdotool` worked for me. I'm only setting forward and back navigation, so it would be great to get an option for this in Cinnamon gestures
2. I missed this and yes, moving the threshold down to the lowest value resulted in my gestures always being picked up. I would agree the default seems far too high, especially for users with a larger touchpad.

**Comment by mtwebster** [comment]
@A-Lovett I'm an idiot - I completely missed the "Execute on" option in the Touche app.

I can definitely fix this, but I'm not sure I'll be able to for the current release. I'll see what we can do.

One-to-one gestures may be achievable with touchegg - I've got some ideas for that.  We'll probably have to move on when we eventually start using Wayland.

**Comment by mtwebster** [comment]
Removed detection:
https://github.com/linuxmint/cinnamon/commit/09b9113171ead0fdf9d0027598b9974f46a8f45b

**Comment by mister-drgn** [comment]
@Starcross Happy to help. You can do a lot more with xdotool like managing windows, moving them between workspaces, etc.

@mtwebster Sounds good, thanks!

**Comment by grepppo** [comment]
@mtwebster I think I my machine would be affected by the changes you have commited to remove the detection, and would be happy to help test it. 
What would be the best way to go about this ?

**Reply by mtwebster** [reply]
Hi, sorry for not responding sooner. If you're willing to edit a couple of files (single line in each) you should be able to have it enabled permanently:

Back up the files first:
```shell
cp /usr/share/cinnamon/cinnamon-settings/modules/cs_gestures.py .
cp /usr/share/cinnamon/js/ui/gestures/gesturesManager.js .
```
Modify cs_gestures.py:
```shell
sudo nano /usr/share/cinnamon/cinnamon-settings/modules/cs_gestures.py
```
Then use the arrow keys down to where is says `DEBUG_SHOW_ALL = False` and  replace `False` with `True` (must be capitalized).

Type <kbd>Ctrl</kbd> + <kbd>S</kbd> then <kbd>Ctrl</kbd> + <kbd>X</kbd> to save and exit.

Now, modify gesturesManager.js:
```shell
sudo nano /usr/share/cinnamon/js/ui/gestures/gesturesManager.js
```
Then use the arrow keys down to where is says `DEBUG_HAVE_DEVICES` and  replace `false` with `true` (must *NOT* be capitalized).

Type <kbd>Ctrl</kbd> + <kbd>S</kbd> then <kbd>Ctrl</kbd> + <kbd>X</kbd> to save and exit.

Cinnamon's gesture settings should now be unlocked, and once you restart Cinnamon (<kbd>Alt</kbd> + <kbd>F2</kbd>, then <kbd>r</kbd>, and <kbd>Enter</kbd>), they should be functional.


*** Note: You don't have to use 'nano' as an editor, but I recommend it as gui editors (like xed), running as root, can end up modifying user files and changing their ownership and/or permissions.

**Comment by HackerKitty1** [comment]
Those of you who have this problem, there is a workaround. 
If you start touchegg by the terminal command, UI for Touche works normally without enabling Gesture.
https://www.reddit.com/r/linuxmint/comments/152zln4/do_i_have_to_remove_touchegg_after_updating_to_212/jsk8cd8/
One annoying issue is that this terminal must be open all time.
I went to back to 21.1 at the moment till a better solution appears.

**Comment by mtwebster** [comment]
fyi: https://github.com/linuxmint/cinnamon/pull/11880

**Reply by igarca** [reply]
The difference is huge. Thank you. Gestures look waaay smoother and faster. 
I would like to suggest adding a _pinch-to-zoom option_ to Gestures. Unfortunately, _xdotool key ctrl+plus_ is not very useful.

**Comment by JosuGZ** [comment]
Hi, I'm interested on having `alt` + `tab` support.

On my previous Linux Mint version I had it, but currently I can't see how to do it (unless I somehow override the config file generated by Mint but I'm not even sure where it is).

Ideally as a generic thing It should be possible to generate events for movement start, movement continues, and end of movement, so we can bind them to keys or commands.

But for that particular task I would appreciate a specific option. It is too weird not to have it.

Particularly on Windows when I drag 3 fingers left and right I switch between windows. They keep changing until I release the gesture and then the last window is the one that stays. On Mint I have something similar but for some reason I can't send more than 15 tab keystrokes.

This is the particular config I had, it is not perfect but it was something:

```xml
    <gesture type="SWIPE" fingers="3" direction="LEFT">
      <action type="SEND_KEYS">
        <repeat>true</repeat>
        <modifiers>Alt_L</modifiers>
        <keys>Shift_L+Tab</keys>
        <decreaseKeys>Tab</decreaseKeys>
      </action>
    </gesture>

    <gesture type="SWIPE" fingers="3" direction="RIGHT">
      <action type="SEND_KEYS">
        <repeat>true</repeat>
        <modifiers>Alt_L</modifiers>
        <keys>Tab</keys>
        <decreaseKeys>Shift_L+Tab</decreaseKeys>
      </action>
    </gesture>
```
